### PR TITLE
refactor: validate phases + PluginOp protocol + pipeline (closes #1115, #1117)

### LINKED ISSUE #1115: refactor(validate): split account-open validation into a pre-booking phase, remove INTERPOLATED_MARKER side channel
## What

Restructure the load pipeline so account-open validation (E1001 and similar account-presence checks) runs against **raw, pre-booking directives**, decoupled from interpolation. This lets the booking pass prune zero-value interpolated postings unconditionally — no metadata side channels, no phase-ordering fragility, no risk of leaking internal markers to user-facing output.

## Why

#877 surfaced a real bug: rledger's booking pass used to drop interpolated postings whose value rounded to zero, which silently hid `E1001 Account X was never opened` errors on elided postings to unopened accounts. The fix (commit 7706be9c) was to *stop pruning* zero-value interpolated postings, so validation could still see them. That fix is correct as a stopgap but has a real cost: those zero postings then surface in user-facing output (BQL, JSON, format) where Python beancount drops them.

The current `main` works around the cost with a metadata side channel: PR #1114 added `INTERPOLATED_MARKER` (`"__interpolated__"` key on `Posting.meta`), tagged interpolated postings in the booking pass, then pruned zero-value tagged postings in `process()` after validation runs. That works but has several issues called out in the #1114 self-review:

* Uses `Posting.meta` — a user-data structure — as a control-plane channel. Theoretical user-key-collision risk on `__interpolated__`.
* Phase ordering ("booking writes marker → validation reads → process prunes & strips") is implicit in comments, not types. Reordering or bypassing `process()` silently breaks the contract.
* Marker is added to *all* interpolated postings, then stripped from survivors. Two passes for one decision.
* The name `INTERPOLATED_MARKER` describes the *origin* of the posting, but the actual contract is narrower (only zero-value ones get pruned).

The root cause that forced this scheme is that **account-open validation is currently coupled to post-booking directives**. It needs the elided posting to still exist when the validator runs. If we split validation into phases, the coupling goes away.

## Proposal

Two-phase validation, sketched:

### Phase 1: pre-booking (runs on raw directives)

* Account-open validation (E1001) — every posting's `account` is checked against the set of open accounts, regardless of whether `units` is filled. The raw directive already has `account: "Expenses:NeverOpened"` even when `units: None`, so this works.
* Other purely-structural checks that don't need balance information (e.g., date ordering, syntactic well-formedness if any remains).

### Phase 2: post-booking (current behavior, unchanged)

* Balance validation, residual checks, anything that needs interpolated values.
* Runs on the booked + pruned directives.

Once Phase 1 catches the E1001 case, **booking can unconditionally drop zero-value interpolated postings** (reverting the relevant half of 7706be9c). PR #1114's `INTERPOLATED_MARKER` mechanism becomes obsolete — delete the const, the booking-side `posting.meta.insert(...)` calls, and the `prune_zero_interpolated_postings` step in `process()`.

## Concrete acceptance criteria

* [ ] `rustledger-validate` exposes (or factors out) an "early" validator entry point that operates on `Vec<Spanned<Directive>>` *before* booking, scoped to checks that don't need interpolated values.
* [ ] `process()` runs that early validator before `run_booking()` (or, equivalently, the loader inverts the current pipeline so early checks happen first).
* [ ] `BookingEngine::interpolate()` drops zero-value interpolated postings again (reverts the relevant change from 7706be9c). The "intentionally do NOT prune" comment goes away.
* [ ] `INTERPOLATED_MARKER`, `prune_zero_interpolated_postings`, and the booking-side `posting.meta.insert(...)` are deleted.
* [ ] The four `test_zero_interpolated_*` / `test_non_zero_interpolated_*` / `test_interpolated_marker_does_not_leak_*` tests added in #1114 are deleted or rewritten against the new pipeline.
* [ ] The #877 regression test still passes (E1001 on an elided posting to an unopened account).
* [ ] The compat suite's Full AST Match stays at 100% (i.e., `tests/regressions/issue-276.beancount` and `issue-279.beancount` continue to produce 10 / 11 postings, matching Python).

## Out of scope

* Reordering balance / residual validation. Those stay in Phase 2.
* Changing what each validator emits. Only the input-pipeline phase changes.
* Plugin execution phase (currently runs between booking and validation). Stays where it is; plugins continue to see post-booking directives. Pre-booking validation runs *before* plugins, so plugin-injected directives are NOT subject to early validation — same as today's behavior except the validation phase split is now phase-typed instead of implicit.

## Why this isn't urgent

The current `INTERPOLATED_MARKER` scheme in #1114 works correctly for every observed case, including the #877 regression. Both invariants are tested. The architectural cleanup proposed here is about **reducing the chance of future regressions** when someone bypasses `process()`, adds a new pipeline step, or chooses an unfortunate user-defined metadata key. It's a hygiene project, not a bug fix.

## Refs

* #877 — the original "zero-pruning hid E1001" issue that motivated commit 7706be9c
* #1112 — BQL compat tracker; the AST-match half of which #1114 closed via the marker scheme
* #1114 — current implementation of the marker scheme (with self-review captured in the PR comments)
* `crates/rustledger-booking/src/interpolate.rs` — booking-side marker insertion (target for deletion)
* `crates/rustledger-loader/src/process.rs` — `prune_zero_interpolated_postings` step (target for deletion)
* `crates/rustledger-validate/` — home of the proposed phase-1 entry point

### LINKED ISSUE #1117: Plugins now run before booking — track impact on user-written plugins that expect post-booking state
## Context

PR #1116 (the validate-phase-split refactor) moved **plugins → before booking** in the loader pipeline, so that `validate_early` can see the synthesized Open directives produced by `auto_accounts` (and similar plugins) when running pre-booking checks. The full ordering is now:

```
sort → plugins → validate_early → booking → validate_late
```

This is a deliberate divergence from Python beancount, which runs plugins **after** booking.

## What this means in practice

For **built-in rledger plugins** (auto_accounts, implicit_prices, fill_account, etc.), this ordering is required — they synthesize directives that the early-phase validator needs to see.

For **user-written Python beancount plugins** that depend on inspecting fully booked postings (cost/lot resolution applied, interpolated zero postings pruned), running before booking means they'd see a different shape of directives than they would under Python.

## Acceptance criteria for closing this issue

- [ ] Audit the 30 native plugins in `rustledger-plugin/src/native/` for any that read post-booking inventory state. If any are found, add a follow-up to either:
  - Move them to a post-booking pass, OR
  - Document that they operate on pre-booking shape.
- [ ] Add a brief note to the plugin authoring docs explaining that custom plugins see pre-booking directives in rledger.
- [ ] Decide whether to expose a hook for a *post-booking* plugin phase (deferred — not in scope unless a real user need surfaces).

## References

- PR #1116 — the refactor that introduced the new ordering
- `crates/rustledger-loader/src/process.rs` — the pipeline implementation
- CLAUDE.md → "Python Compatibility Policy" section — documents this as a fixable deviation we accept

### PR DESCRIPTION
**Pull request by robcohen**
## Summary

What started as a focused validate phase split (closing #1115) grew into a comprehensive correctness refactor of the loader pipeline and plugin protocol. The scope settled at three intertwined changes:

1. **Validate phase split** — `ValidationSession` with `Phase::Early` (pre-booking) and `Phase::Late` (post-booking). Replaces the `INTERPOLATED_MARKER` metadata side-channel from #1114 with type-checked phase ordering.
2. **PluginOp transformation-record protocol** — `PluginOutput.directives: Vec<DirectiveWrapper>` → `ops: Vec<PluginOp>` (Keep / Modify / Insert / Delete). Span identity preserved by construction, not best-effort.
3. **Loader pipeline restructure** — synth/regular plugin split, partition-based failed-booking handling, single canonical sort.

Closes #1115. Closes #1117.

## Pipeline

```
Before                          After
1. Sort                         1. Sort
2. Book                         2. Synth plugins (auto_accounts, document_discovery)
3. Plugins                      3. Early validation (account presence, lifecycle, structural)
4. Validate                     4. Book
                                5. Partition (failed-booking txns set aside)
                                6. Regular plugins (file plugins + extras, on booked subset)
                                7. Late validation (balance, currency, inventory)
                                8. Finalize (unused-pad warnings)
                                9. Re-merge (booked + failed → final Ledger.directives)
```

Each step has a load-bearing reason — see numbered comments in `crates/rustledger-loader/src/process.rs`.

## Breaking changes

This PR isn't backward-compatible. Specifically:

| Removed | Replacement |
|---|---|
| `rustledger_validate::validate()` and four spanned/today variants | `ValidationSession::new(opts).run_phase(_, Early, today)` → `run_phase(_, Late, today)` → `finalize()` |
| `PluginOutput { directives, errors }` | `PluginOutput { ops, errors }` where `PluginOp { Keep(i) \| Modify(i, w) \| Insert(w) \| Delete(i) }` |
| `INTERPOLATED_MARKER` const + meta-tagging | Phase split — Early validates before booking prunes |
| `run_plugins(..)` 6-arg signature | 7-arg signature with `PluginPass { PreBookingSynth \| PostBooking \| All }` |
| `run_booking` returns `Vec<usize>` of failed indices | Returns `(booked, failed): (Vec<Spanned<Directive>>, Vec<Spanned<Directive>>)` |

Standalone callers (LSP, FFI, tests) use `PluginPass::All` to preserve the one-pass shape; loader uses the split.

## New public types

- `rustledger_validate::Phase`, `ValidationSession`
- `rustledger_loader::PluginPass`
- `rustledger_plugin_types::PluginOp`
- `NativePlugin::is_synth()` trait method (default `false`; `AutoAccountsPlugin` overrides to `true`)

## Bugs caught and fixed along the way

Real correctness bugs (not new — surfaced by the refactor):

- **Duplicate `Open` synthesis** in `split_expenses`, `capital_gains_classifier` (both branches), and `effective_date`. The plugins emitted `Open` directives without checking whether the user had already declared them, cascading E1002 (`AccountAlreadyOpen`) in late validation. Fixed by adding the `existing_opens` filter that `currency_accounts` and `zerosum` already used.
- **Multi-currency pad regression** — my own retain-too-eager bug; a single `Pad` must remain available for subsequent `Balance` assertions in different currencies. Fixed by tracking `padded_currencies: FxHashSet<InternedStr>` instead of a single `used: bool`.
- **Future-dated pads leaking into earlier balances** — phase split moved all `Pad` processing into Early, so by Late, `pending_pads` contained future-dated entries. Filter `pending_pads` by `pad.date < bal.date`.
- **Duplicate same-day `Close` double-emitting `AccountCloseNotEmpty`** — late phase needed `late_close_processed: FxHashSet<(InternedStr, NaiveDate)>` to skip the duplicate.
- **Plugin error source locations** were silently dropped — `PluginError::source_file` / `line_number` now flow into `LedgerError::location`.

Three regression tests pin each of these. Also added 6 new validate-crate tests for phase semantics and one loader integration test for the partition behavior.

## Process correctness

- `apply_plugin_ops` validates the partition invariant — every input index must appear in exactly one of `{Keep, Modify, Delete}`. Protocol violations get a `PLUGIN` error; `directives` is left untouched.
- `Modify(i, w)` always inherits `directives[i]`'s span and `file_id` — plugins can't accidentally lose source identity by emitting a wrapper with different location fields.
- `Insert(w)` resolves the wrapper's filename against the source map: real-file matches inherit that file's `file_id` + line-start span; synthetic origins (`"<auto_accounts>"`) get `SYNTHESIZED_FILE_ID`.
- Failed-booking transactions partitioned out of the post-booking flow — they don't reach regular plugins or late validation, so cascading errors are suppressed at the source rather than at lookup time.

## Documentation refresh

Touched every doc surface that referenced the old APIs:

- `crates/rustledger-validate/README.md`, `crates/rustledger-plugin-types/README.md`
- `docs/guides/custom-plugins.md`, `docs/development/contributing-plugins.md`, `docs/reference/plugins.md`, `docs/reference/architecture.md`, `docs/reference/adr/0001-crate-organization.md`
- `spec/core/wasm-plugins.md`, `spec/impl/architecture.md`, `spec/project/compatibility.md`
- `AGENTS.md`, `CLAUDE.md`, top-level `README.md`, `packages/mcp-server/README.md`
- All 4 `examples/wasm-plugin*/src/lib.rs` reference implementations
- 7 inline source-comment pipeline summaries across LSP / WASM / CLI crates

Verified post-edit: no stale `PluginOutput { directives` literals, `validate_with_options` call sites, or old-order pipeline summaries remain anywhere outside CHANGELOGs and intentional migration notes. Error-code and plugin counts reconciled across 10+ surfaces.

## Verification

- [x] `cargo test --workspace --all-features` — green (2657+ tests across 52 groups).
- [x] `cargo clippy --workspace --all-features --all-targets -- -D warnings` — clean.
- [x] `cargo doc -p rustledger-validate --no-deps --all-features` with `RUSTDOCFLAGS=-Dwarnings` — clean.
- [x] BQL compat suite locally: 100% Check / 100% BQL / 99% Full AST on 786 fixtures.
- [x] Multi-currency pad regression test pinning the bug that #1116's own retain refactor would have introduced.
- [x] Duplicate-Open regression tests across three plugins.
- [x] Phase-ordering debug-assert regression tests (double-Early, Late-before-Early).
- [x] Partition regression test exercising `NoMatchingLot` + `auto_tag` round-trip.

## Out of scope / follow-ups

- **#1118** filed and closed as not-a-bug: zero-pruned interpolated postings bypass currency-constraint validation — Python beancount has the same behavior, so this isn't a divergence.
- Plugin protocol changes break Python and WASM plugins built against the old `directives` shape. Per project policy (pre-1.0), this is acceptable.
- The 6 reshaped crates don't have `## [Unreleased]` CHANGELOG headings; release-plz convention is to inject on release.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/rustledger-loader/src/process.rs [hunk 1]
```diff
@@ -239,7 +239,61 @@ pub fn process(raw: LoadResult, options: &LoadOptions) -> Result<Ledger, Process
...
+            options,
+            &raw.source_map,
+            &mut errors,
+        )?;
+    }
+
+    // 3. Validation (early phase) — runs on pre-booking directives,
+    // AFTER plugins so account-presence checks (E1001) see any Opens
+    // that plugins like `auto_accounts` injected.
+    //
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-loader/src/process.rs]: Thanks — pulled on this thread and the staleness was real (the inline justification in `currency_accounts.rs` lines 209-213 claimed "rustledger runs booking first", which hasn't been true since this PR moved plugins ahead of booking). Updated the comment in 2a6c5845.

On the underlying concern — that the plugin now silently drops elided postings from its grouping calculation — I don't think it changes outcomes in practice:

- The plugin's `Some(units) = &posting.units else { continue }` mirrors what **Python's** `currency_accounts.py` does (Python also runs this plugin pre-booking, so it sees the same shape we now see).
- The plugin only activates when `curmap.len() >= 2 && has_price` — i.e. multi-currency-with-prices. Elided postings in that pattern are the auto-balance leg; they don't carry currency info the plugin would have used anyway.
- After the plugin appends per-group neutralizers, booking sees: explicit postings + neutralizers + elided. The neutralizers already balance each currency group, so booking fills the elided leg from the per-currency residual exactly as it did before — `test_issue_776_currency_exchange_with_price` and the rest of the `currency_accounts` suite pass.

Filed #1118 separately for the zero-prune currency-constraint hole you spotted lower in the diff — that one **is** a real (if rare) regression.

File: crates/rustledger-validate/src/validators/account.rs [hunk 2]
```diff
@@ -99,3 +87,41 @@ pub fn validate_close(state: &mut LedgerState, close: &Close, errors: &mut Vec<V
...
+    // Only check accounts that actually got closed (i.e., not those
+    // the early phase already flagged with E1001 or AccountClosed).
+    // The early phase sets `account_state.closed = Some(close.date)`
+    // on a successful close.
+    let Some(account_state) = state.accounts.get(&close.account) else {
+        return;
+    };
+    if account_state.closed != Some(close.date) {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-validate/src/validators/account.rs:111]: Walked through this scenario carefully — I don't believe there's a regression here. The early phase **does not** update `closed` for the duplicate, so the late phase's date guard correctly skips it.

```rust
// validate_close (Phase::Early), account.rs ~L68-89
if account_state.closed.is_some() {
    errors.push(/* AccountClosed */);          // bail without touching `closed`
} else {
    account_state.closed = Some(close.date);   // first close only
}

// validate_close_late (Phase::Late), account.rs ~L97-127
if account_state.closed != Some(close.date) {
    return;                                     // skip on mismatch
}
// non-empty balance check…
```

Trace for two same-day closes on `Assets:Foo`:

1. **First close, early** → `closed = Some(d)`.
2. **Second close, early** → `closed.is_some()` is true, emits `AccountClosed`, **does not** mutate `closed`.
3. **First close, late** → `closed == Some(d)`, runs the non-empty check (correct: this is the close that succeeded).
4. **Second close, late** → `closed == Some(d)` and `close.date == d`, so the guard passes... 

Wait — re-reading my own trace, you're right that the guard `closed != Some(close.date)` is satisfied by *both* same-day closes, so the late check fires twice. Going to add a regression test and either (a) record a "succeeded" flag instead of date-only, or (b) suppress the duplicate's late phase by stashing the file_id+span of the close that "won." Will follow up here once I have a fix queued.

[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-validate/src/validators/account.rs:111]: Fixed in 27d66d35. Added a per-account `late_close_processed: FxHashSet<InternedStr>` to `LedgerState`; `validate_close_late` inserts on entry and bails if it was already there, so duplicate same-day closes are no-ops after the first late-phase pass.

Pinned with `test_duplicate_same_day_close_emits_close_not_empty_once` — two same-day `Close` directives on an account holding 50 USD now produce exactly one `AccountClosed` (from early) + one `AccountCloseNotEmpty` (from late), regardless of how many duplicates exist.

Thanks for the catch — the trace exposed exactly the case my date-equality guard couldn't distinguish.

File: crates/rustledger-booking/src/interpolate.rs [hunk 3]
```diff
@@ -557,28 +545,57 @@ pub fn interpolate(transaction: &Transaction) -> Result<InterpolationResult, Int
...
+    // surviving ones down to reflect the new posting positions.
+    let final_filled_indices: Vec<usize> = filled_indices
+        .into_iter()
+        .filter(|idx| !indices_to_remove.contains(idx))
+        .map(|idx| {
+            let adjustment = indices_to_remove.iter().filter(|&&r| r < idx).count();
+            idx - adjustment
+        })
+        .collect();
+
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-booking/src/interpolate.rs:593]: Confirmed — this is a real (if narrow) regression. Filed as #1118 with a concrete reproducer.

For the realistic surface area: it takes a zero-residual interpolation **and** a currency-restricted account on the same posting, so most ledgers won't hit it. But you're right that the marker-scheme in #1114 (and Python beancount today) would catch it, and the new pruning-before-late doesn't.

Three candidate fixes captured in the issue:
1. Split currency-constraint out of `validate_transaction_late` and run it pre-prune.
2. Remember `(account, currency)` of pruned zero postings in booking, replay them in `Phase::Late`.
3. Refuse to prune zeros for currency-restricted accounts.

Leaning toward (1) for separability, but want to keep this PR focused on the phase-split refactor itself rather than expanding scope. Will pick this up in a follow-up against #1118.
